src/core/service.c

```diff
@@ -3059,6 +3060,16 @@ static void service_notify_message(Unit *u, pid_t pid, char **tags, FDSet *fds)
                 else
                         log_unit_debug(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID which is currently not known", pid);
                 return;
+        } else if (s->notify_access == NOTIFY_EXEC && pid != s->main_pid && pid != s->control_pid) {
+                if (s->main_pid != 0 && s->control_pid != 0)
+                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID "PID_FMT" and control PID "PID_FMT,
+                                         pid, s->main_pid, s->control_pid);
+                else if (s->main_pid != 0 || s->control_pid != 0)
+                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for %s PID "PID_FMT,
+                                         pid, s->main_pid ? "main" : "control", s->main_pid | s->control_pid);
```

Comment:
Ok, but then the ternary operator for "main" vs. "control" has to be changed too and the part becomes
```s->main_pid != 0 ? "main" : "control", s->main_pid != 0 ? s->main_pid : s->control_pid```
This almost warrants a separate if block.

The alternative
```s->main_pid ? "main" : "control", s->main_pid ?: s->control_pid```
looks pretty neat if you ask me.

In the end I can see the appeal in either. Shall I add a clause to the if block?
```
                if (s->main_pid != 0 && s->control_pid != 0)
                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID "PID_FMT" and control PID "PID_FMT,
                                         pid, s->main_pid, s->control_pid);
                else if (s->main_pid != 0)
                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID "PID_FMT, pid, s->main_pid);
                else if (s->control_pid != 0)
                        log_unit_warning(u, "Got notification message from PID "PID_FMT", but reception only permitted for control PID "PID_FMT, pid, s->control_pid);
                else
                        log_unit_debug(u, "Got notification message from PID "PID_FMT", but reception only permitted for main PID and control PID which are currently not known", pid);
                return;
```

P.S. Do we really want the last case to be a debug message?